# Bank Marketing

The <b>bank-marketing.csv</b> data is related with direct marketing campaigns of a Portuguese banking institution. The marketing campaigns were based on phone calls. Often, more than one contact to the same client was required, in order to access if the product (bank term deposit) would be (or not) subscribed. The ultimate goal is to predict if the client will subscribe to a term deposit (variable y). This is a classic classification problem where the attempt is to classify between two classes - those who'll subscribe and those who won't.

Dataset reference:
- S. Moro, R. Laureano and P. Cortez. Using Data Mining for Bank Direct Marketing: An Application of the CRISP-DM Methodology. 
- In P. Novais et al. (Eds.), Proceedings of the European Simulation and Modelling Conference - ESM'2011, pp. 117-121, Guimarães, Portugal, October, 2011. EUROSIS.

#### Variable description:

- 1 age (numeric)
- 2 job : type of job (categorical: "admin.","unknown","unemployed","management","housemaid","entrepreneur","student", "blue-collar","self-employed","retired","technician","services") 
- 3 marital : marital status (categorical: "married","divorced","single"; note: "divorced" means divorced or widowed)
- 4 education (categorical: "unknown","secondary","primary","tertiary")
- 5 default: has credit in default? (binary: "yes","no")
- 6 balance: average yearly balance, in euros (numeric) 
- 7 housing: has housing loan? (binary: "yes","no")
- 8 loan: has personal loan? (binary: "yes","no")
   
#### related with the last contact of the current campaign:
- 9 contact: contact communication type (categorical: "unknown","telephone","cellular") 
- 10 day: last contact day of the month (numeric)
- 11 month: last contact month of year (categorical: "jan", "feb", "mar", ..., "nov", "dec")
- 12 duration: last contact duration, in seconds (numeric)

#### other attributes:
- 13 campaign: number of contacts performed during this campaign and for this client (numeric, includes last contact)
- 14 pdays: number of days that passed by after the client was last contacted from a previous campaign (numeric, -1 means client was not previously contacted)
- 15 previous: number of contacts performed before this campaign and for this client (numeric)
- 16 poutcome: outcome of the previous marketing campaign (categorical: "unknown","other","failure","success")

Output variable (desired target):
- 17 y - has the client subscribed a term deposit? (binary: "yes","no")

### Read the dataset and answer the following questions.

In [1]:
import numpy as np

In [2]:
import pandas as pd

In [3]:
df = pd.read_csv('bank-marketing.csv')

In [4]:
df

age            job  marital  education default  balance housing loan  \
0      30     unemployed  married    primary      no     1787      no   no   
1      33       services  married  secondary      no     4789     yes  yes   
2      35     management   single   tertiary      no     1350     yes   no   
3      30     management  married   tertiary      no     1476     yes  yes   
4      59    blue-collar  married  secondary      no        0     yes   no   
...   ...            ...      ...        ...     ...      ...     ...  ...   
4516   33       services  married  secondary      no     -333     yes   no   
4517   57  self-employed  married   tertiary     yes    -3313     yes  yes   
4518   57     technician  married  secondary      no      295      no   no   
4519   28    blue-collar  married  secondary      no     1137      no   no   
4520   44   entrepreneur   single   tertiary      no     1136     yes  yes   

       contact  day month  duration  campaign  pdays  previous poutcome   y  
0     cellular   19   oct        79         1     -1         0  unknown  no  
1     cellular   11   may       220         1    339         4  failure  no  
2     cellular   16   apr       185         1    330         1  failure  no  
3      unknown    3   jun       199         4     -1         0  unknown  no  
4      unknown    5   may       226         1     -1         0  unknown  no  
...        ...  ...   ...       ...       ...    ...       ...      ...  ..  
4516  cellular   30   jul       329         5     -1         0  unknown  no  
4517   unknown    9   may       153         1     -1         0  unknown  no  
4518  cellular   19   aug       151        11     -1         0  unknown  no  
4519  cellular    6   feb       129         4    211         3    other  no  
4520  cellular    3   apr       345         2    249         7    other  no  

[4521 rows x 17 columns]

### Question 1

Extract all column names. Count the number of columns (using code).

In [5]:
df.columns

Index(['age', 'job', 'marital', 'education', 'default', 'balance', 'housing',
       'loan', 'contact', 'day', 'month', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'y'],
      dtype='object')

In [6]:
len(df.columns)

17

### Question 2

Is data in the correct format? By that we mean, do you see integers and floats where you expect them to be the case? If not, convert them into the correct format. Also make sure that all entries are non-null.

<b>Answer: The column 'balance' has the data type int and has been converted into float.</b>

In [7]:
df.dtypes

age           int64
job          object
marital      object
education    object
default      object
balance       int64
housing      object
loan         object
contact      object
day           int64
month        object
duration      int64
campaign      int64
pdays         int64
previous      int64
poutcome     object
y            object
dtype: object

In [8]:
df['balance'] = df['balance'].astype(float)

In [9]:
df.dtypes

age            int64
job           object
marital       object
education     object
default       object
balance      float64
housing       object
loan          object
contact       object
day            int64
month         object
duration       int64
campaign       int64
pdays          int64
previous       int64
poutcome      object
y             object
dtype: object

In [10]:
df.age.values 

array([30, 33, 35, ..., 57, 28, 44], dtype=int64)

In [11]:
df[[ (not a.dtype == 'int64') |  (a is None) for a in df.age.values ]]

Empty DataFrame
Columns: [age, job, marital, education, default, balance, housing, loan, contact, day, month, duration, campaign, pdays, previous, poutcome, y]
Index: []

In [12]:
df[[ (a is None) for a in df.day.values ]]

Empty DataFrame
Columns: [age, job, marital, education, default, balance, housing, loan, contact, day, month, duration, campaign, pdays, previous, poutcome, y]
Index: []

In [13]:
df[[ (not a.dtype == 'float') | (a is None) for a in df.balance.values ]]

Empty DataFrame
Columns: [age, job, marital, education, default, balance, housing, loan, contact, day, month, duration, campaign, pdays, previous, poutcome, y]
Index: []

In [14]:
df[[ (not a.dtype == 'int64') | (a is None) for a in df.day.values ]]

Empty DataFrame
Columns: [age, job, marital, education, default, balance, housing, loan, contact, day, month, duration, campaign, pdays, previous, poutcome, y]
Index: []

In [15]:
df[[ (not a.dtype == 'int64') | (a is None) for a in df.duration.values ]]

Empty DataFrame
Columns: [age, job, marital, education, default, balance, housing, loan, contact, day, month, duration, campaign, pdays, previous, poutcome, y]
Index: []

In [16]:
df[[ (not a.dtype == 'int64') | (a is None) for a in df.campaign.values ]]

Empty DataFrame
Columns: [age, job, marital, education, default, balance, housing, loan, contact, day, month, duration, campaign, pdays, previous, poutcome, y]
Index: []

In [17]:
df[[ (not a.dtype == 'int64') | (a is None) for a in df.pdays.values ]]

Empty DataFrame
Columns: [age, job, marital, education, default, balance, housing, loan, contact, day, month, duration, campaign, pdays, previous, poutcome, y]
Index: []

In [18]:
df[[ (not a.dtype == 'int64') | (a is None) for a in df.previous.values ]]

Empty DataFrame
Columns: [age, job, marital, education, default, balance, housing, loan, contact, day, month, duration, campaign, pdays, previous, poutcome, y]
Index: []

<b>Answer: There did not return an entry where there is null or inconsistent data type.</b>

### Question 3

Provide a general summary statistics of the entire dataset. The describe() method is what you would want to use.

In [19]:
df.describe()

age       balance          day     duration     campaign  \
count  4521.000000   4521.000000  4521.000000  4521.000000  4521.000000   
mean     41.170095   1422.657819    15.915284   263.961292     2.793630   
std      10.576211   3009.638142     8.247667   259.856633     3.109807   
min      19.000000  -3313.000000     1.000000     4.000000     1.000000   
25%      33.000000     69.000000     9.000000   104.000000     1.000000   
50%      39.000000    444.000000    16.000000   185.000000     2.000000   
75%      49.000000   1480.000000    21.000000   329.000000     3.000000   
max      87.000000  71188.000000    31.000000  3025.000000    50.000000   

             pdays     previous  
count  4521.000000  4521.000000  
mean     39.766645     0.542579  
std     100.121124     1.693562  
min      -1.000000     0.000000  
25%      -1.000000     0.000000  
50%      -1.000000     0.000000  
75%      -1.000000     0.000000  
max     871.000000    25.000000

### Question 4

The data type of columns like job, married, education etc. is called categorical data. List all the different categories in the job column.

In [20]:
df.job.values

array(['unemployed', 'services', 'management', ..., 'technician',
       'blue-collar', 'entrepreneur'], dtype=object)

In [21]:
df.loc[~df.job.duplicated(), :]

age            job  marital  education default  balance housing loan  \
0    30     unemployed  married    primary      no   1787.0      no   no   
1    33       services  married  secondary      no   4789.0     yes  yes   
2    35     management   single   tertiary      no   1350.0     yes   no   
4    59    blue-collar  married  secondary      no      0.0     yes   no   
6    36  self-employed  married   tertiary      no    307.0     yes   no   
7    39     technician  married  secondary      no    147.0     yes   no   
8    41   entrepreneur  married   tertiary      no    221.0     yes   no   
11   43          admin  married  secondary      no    264.0     yes   no   
13   20        student   single  secondary      no    502.0      no   no   
24   26      housemaid  married   tertiary      no    543.0      no   no   
27   67        retired  married    unknown      no    696.0      no   no   
95   41        unknown   single   tertiary      no   1567.0      no   no   

      contact  day month  duration  campaign  pdays  previous poutcome    y  
0    cellular   19   oct        79         1     -1         0  unknown   no  
1    cellular   11   may       220         1    339         4  failure   no  
2    cellular   16   apr       185         1    330         1  failure   no  
4     unknown    5   may       226         1     -1         0  unknown   no  
6    cellular   14   may       341         1    330         2    other   no  
7    cellular    6   may       151         2     -1         0  unknown   no  
8     unknown   14   may        57         2     -1         0  unknown   no  
11   cellular   17   apr       113         2     -1         0  unknown   no  
13   cellular   30   apr       261         1     -1         0  unknown  yes  
24   cellular   30   jan       169         3     -1         0  unknown   no  
27  telephone   17   aug       119         1    105         2  failure   no  
95   cellular    1   jul       291         2     -1         0  unknown   no

In [22]:
df.job.drop_duplicates()

0        unemployed
1          services
2        management
4       blue-collar
6     self-employed
7        technician
8      entrepreneur
11            admin
13          student
24        housemaid
27          retired
95          unknown
Name: job, dtype: object

### Question 5

In one line of code, provide a count of the number of people who were unemployed and owned a home.

In [23]:
len(df.loc[(df.job == 'unemployed') & (df.housing == 'yes'), :])

58

### Question 6

What is the education level of a typical blue-collar worker? Explore value_counts() method.

In [24]:
df.education.drop_duplicates()

0       primary
1     secondary
2      tertiary
20      unknown
Name: education, dtype: object

In [25]:
df.loc[df.job == 'blue-collar', 'education'].value_counts()

secondary    524
primary      369
unknown       41
tertiary      12
Name: education, dtype: int64

### Question 7

How many who are unemployed have an outstanding loan? Is that percentage more than that of the employed ones? 

<b>For Answer 1: Assume only a personal loan is considered as an outstanding loan and "employed" is not "unemployed"</b>

<b># of ppl who are unemployed have an outstadning loan (personal):</b>

In [66]:
len(df.loc[(df.job == 'unemployed') & (df.loan == 'yes') , :])

13

<b># of ppl who are unemployed</b>

In [27]:
len(df.loc[(df.job == 'unemployed'), :])

128

<b>Percentage of ppl who are unemployed and have an outstanding loan (personal):</b>

In [67]:
len(df.loc[(df.job == 'unemployed') & (df.loan == 'yes'), :])/len(df.loc[(df.job == 'unemployed'), :])*100

10.15625

<b># of ppl who are employed have an outstanding loan (assuming employed is non-unemployed):</b>

In [68]:
len(df.loc[~(df.job == 'unemployed') & (df.loan == 'yes' ), :])

678

<b># of ppl who are employed (assuming employed is non-unemployed):</b>

In [30]:
len(df.loc[~(df.job == 'unemployed'), :])

4393

<b>Percentage of ppl who are employed and have an outstanding loan (assuming employed is non-unemployed):</b>

In [69]:
len(df.loc[~(df.job == 'unemployed') &  (df.loan == 'yes') , :]) / len(df.loc[~(df.job == 'unemployed'), :]) * 100

15.43364443432734

<b>Answer 1: Percent who are unemployed have an outstanding loan is 10.16% while percent who are employed have an outstanding loan is 15.43% (assuming employed is non-unemployed). Hence that percent of unemployed is less than that of employed.</b>  

<b>For Answer 2: Assume either a housing loan or a personal loan is considered as an outstanding loan and "employed" can be 2a) non-unemployed or 2b) can exclude "student", "retired" and "unknown".</b>

<b># of ppl who are unemployed have an outstanding loan (housing + personal):</b>

In [26]:
len(df.loc[(df.job == 'unemployed') & ( (df.housing == 'yes') | (df.loan == 'yes') ), :])

62

<b># of ppl who are unemployed</b>

In [27]:
len(df.loc[(df.job == 'unemployed'), :])

128

<b>Percentage of ppl who are unemployed and have an outstanding loan (housing + personal):</b>

In [28]:
len(df.loc[(df.job == 'unemployed') & ( (df.housing == 'yes') | (df.loan == 'yes') ), :])/len(df.loc[(df.job == 'unemployed'), :])*100

48.4375

<b>2a) # of ppl who are employed have an outstanding loan (assuming employed is non-unemployed):</b>

In [29]:
len(df.loc[~(df.job == 'unemployed') & ( (df.housing == 'yes') | (df.loan == 'yes') ), :])

2782

<b>2a) # of ppl who are employed (assuming employed is non-unemployed):</b>

In [30]:
len(df.loc[~(df.job == 'unemployed'), :])

4393

<b>2a) Percentage of ppl who are employed and have an outstanding loan (assuming employed is non-unemployed):</b>

In [31]:
len(df.loc[~(df.job == 'unemployed') & ( (df.housing == 'yes') | (df.loan == 'yes') ), :]) / len(df.loc[~(df.job == 'unemployed'), :]) * 100

63.32802185294787

<b>2b) # of ppl who are employed have an outstanding loan (assuming employed excludes student, retired, unknown):</b>

In [32]:
len(df[ [(s not  in ['student', 'retired','unknown', 'unemployed']) for s in df.job.values] & ( (df.housing == 'yes') | (df.loan == 'yes') )])

2691

<b>2b) # of ppl who are employed (assuming employed excludes student, retired, unknown):</b>

In [33]:
len(df[ [(s not  in ['student', 'retired','unknown', 'unemployed']) for s in df.job.values] ])

4041

<b>2b) Percentage of ppl who are employed and have an outstanding loan (assuming employed excludes student, retired, unknown):</b>

In [34]:
len(df[ [(s not  in ['student', 'retired','unknown', 'unemployed']) for s in df.job.values] & ( (df.housing == 'yes') | (df.loan == 'yes') )]) / len(df[ [(s not  in ['student', 'retired','unknown', 'unemployed']) for s in df.job.values] ]) * 100

66.59242761692651

<b>Answer 2: Percent who are unemployed have an outstanding loan is 48.44% while percent who are employed have an outstanding loan is 2a) 63.32% (assuming employed is non-unemployed) or 2b) 66.59% (assuming employed excludes student, retired, unknown). Hence that percent of unemployed is less than that of employed regardless which way the employed is assumed.</b>  

### Question 8

What percent of clients subscribed to the term deposit (column y)? 

In [35]:
len(df.loc[(df.y == 'yes'), :]) / len(df) * 100

11.523999115239992

### Question 9

From all the clients, what percent of married clients subscribed to the term deposit? Is that more or less than that for single folks?

<b>Percent of married clients subscribed to the term deposit:</b>

In [36]:
len(df.loc[((df.marital == 'married') & (df.y == 'yes')), :]) / len(df) * 100

6.126963061269631

<b>Percent of single clients subscribed to the term deposit:</b>

In [37]:
len(df.loc[((df.marital == 'single') & (df.y == 'yes')), :]) / len(df) * 100

3.6938730369387303

<b>Answer: 6.13% of married clients subscribed to the term despoit while only 3.69% of single clients subscribed. Hence, more married clients subscribed than that of singles.</b>

### Question 10

<p>Ask an interesting question of this data set and provide a solution to answer that.</p>
<p>The goal is to help teach your fellow students all possible questions we can collectively ask of this dataset. Your question should be as clear as possible and as short as possible. Try to avoid asking questions that are too trivial or obvious.</p>

<p>This is a bonus point question. The only way not to get full points are if you do the following:
<ul>
<li>You do not perform this task.
<li>You do not include a solution. 
</ul>
</p>
<p>If the solution you provide is incorrect, you will still receive full points. But you must make an honest effort to get it right.
</p>

Write your question here.

<b>My Question: We want to acheive higher success rate in the next campaign. From those who subscribed to the term deposit, find out how they are contacted, what kind of duration and how often they are contacted that made up more than 70% of eventually subscribed clients?</b>  

In [38]:
len(df.loc[(df.y == 'yes'), :])

521

In [40]:
df.loc[(df.y == 'yes'), ['contact', 'duration','campaign', 'y']]

contact  duration  campaign    y
13     cellular       261         1  yes
30    telephone       897         2  yes
33     cellular       958         6  yes
34     cellular       354         3  yes
36    telephone        97         1  yes
...         ...       ...       ...  ...
4494    unknown       576         3  yes
4503   cellular       816         6  yes
4504   cellular       951         3  yes
4505    unknown      1234         3  yes
4511    unknown      1263         2  yes

[521 rows x 4 columns]

In [41]:
df.loc[df.y == 'yes', 'contact'].value_counts()

cellular     416
unknown       61
telephone     44
Name: contact, dtype: int64

In [42]:
len(df.loc[(df.y == 'yes') & (df.contact == 'cellular'),:]) / len(df.loc[(df.y == 'yes'), :]) *100

79.84644913627639

<b>79% of clients subscribed to the term desposit are contacted by cellular phone.</b>

In [43]:
df.loc[df.y == 'yes', 'duration'].sort_values()

3652      30
4124      76
1727      78
1467      80
3874      87
        ... 
2827    1971
3853    2029
980     2087
4123    2456
3673    2769
Name: duration, Length: 521, dtype: int64

In [44]:
df.loc[(df.y == 'yes'), 'duration'].describe()

count     521.000000
mean      552.742802
std       390.325805
min        30.000000
25%       260.000000
50%       442.000000
75%       755.000000
max      2769.000000
Name: duration, dtype: float64

In [45]:
len(df.loc[(df.y == 'yes') & (df.duration <= 100), :])

15

In [46]:
len(df.loc[(df.y == 'yes')  & (df.duration > 100) & (df.duration <= 200), :])

54

In [47]:
len(df.loc[(df.y == 'yes')  & (df.duration > 200) & (df.duration <= 300), :])

96

In [48]:
len(df.loc[(df.y == 'yes')  & (df.duration > 300) & (df.duration <= 400), :])

70

In [49]:
len(df.loc[(df.y == 'yes')  & (df.duration > 400) & (df.duration <= 500), :])

56

In [50]:
len(df.loc[(df.y == 'yes')  & (df.duration > 500) & (df.duration <= 600), :])

40

In [51]:
len(df.loc[(df.y == 'yes')  & (df.duration > 600) & (df.duration <= 700), :])

40

In [52]:
len(df.loc[(df.y == 'yes')  & (df.duration > 700) & (df.duration <= 800), :])

38

In [53]:
len(df.loc[(df.y == 'yes')  & (df.duration > 800) & (df.duration <= 900), :])

30

In [54]:
len(df.loc[(df.y == 'yes')  & (df.duration > 900) & (df.duration <= 1000), :])

21

In [55]:
len(df.loc[(df.y == 'yes')  & (df.duration > 1000) & (df.duration <= 2000), :])

57

In [56]:
len(df.loc[(df.y == 'yes')  & (df.duration > 2000), :])

4

In [57]:
len(df.loc[(df.y == 'yes')  & (df.duration > 200) & (df.duration <= 600), :])

262

In [58]:
len(df.loc[(df.y == 'yes')  & (df.duration > 200) & (df.duration <= 600), :]) / len(df.loc[(df.y == 'yes'), :]) * 100

50.287907869481764

<b>50% of clients have been contacted for between 200 and 600 minutes. Most clients have been contacted for between 200 and 300 minutes.</b>

In [59]:
df.loc[df.y == 'yes', 'campaign'].value_counts()

1     240
2     138
3      57
4      43
6      16
5       9
7       6
8       4
13      2
9       2
24      1
17      1
12      1
10      1
Name: campaign, dtype: int64

In [60]:
len(df.loc[df.y == 'yes', 'campaign'])

521

In [61]:
df.loc[df.y == 'yes', 'campaign'].describe()

count    521.000000
mean       2.266795
std        2.092071
min        1.000000
25%        1.000000
50%        2.000000
75%        3.000000
max       24.000000
Name: campaign, dtype: float64

In [62]:
len(df.loc[((df.y == 'yes') & (df.campaign == 1)), :]) / len(df.loc[(df.y == 'yes'),:]) * 100

46.06525911708253

In [63]:
len(df.loc[(df.y == 'yes') & ((df.campaign == 1)  | (df.campaign == 2)), :]) / len(df.loc[(df.y == 'yes'),:]) * 100

72.552783109405

<b>46% of clients who subscribed to term desposit have been contacted for 1 time. 72% of clients who subscribed have been contacted for 1 or 2 times. There seems most clients who subscribed have been contacted between 1 and 2 times. There does not seem contacting more frequent may result more subscription to term desposit.</b> 

<b>Based on the statistics, most clients who subscribed the term desposit were contacted by cellular, for duration between 200 and 500, and for 1 or 2 times to achieve more than 70% of successful rate. </b>

<b>Note: There has not taken the demography or other stats into analysis. Hence, the success rate can be caused by or correlated to other factors by itself or combined. </b>